packages/ketcher-polymer-editor-react/src/components/modal/Open/Open.tsx

```diff
@@ -18,39 +18,113 @@ import { useCallback, useEffect, useState } from 'react';
+import {
+  ChemicalMimeType,
+  KetSerializer,
+  StructService,
+  SupportedFormat,
+  identifyStructFormat,
+  CoreEditor,
+  KetcherLogger,
+} from 'ketcher-core';
+import { IndigoProvider } from 'ketcher-react';
 import assert from 'assert';
 import { RequiredModalProps } from '../modalContainer';
+import { OpenFileWrapper } from './Open.styles';
+import { Loader } from '../save/Save.styles';
+import { LoadingCircles } from './AnalyzingFile/LoadingCircles';
+import { useAppDispatch } from 'hooks';
+import { openErrorModal } from 'state/modal';
+import { AnyAction, Dispatch } from 'redux';
 
-const MODAL_STATES = {
+export interface Props {
+  onClose: () => void;
+  isModalOpen: boolean;
+}
+
+export const MODAL_STATES = {
   openOptions: 'openOptions',
   textEditor: 'textEditor',
+} as const;
+
+export type MODAL_STATES_VALUES =
+  typeof MODAL_STATES[keyof typeof MODAL_STATES];
+
+const addToCanvas = ({
+  ketSerializer,
+  editor,
+  struct,
+}: {
+  ketSerializer: KetSerializer;
+  editor: CoreEditor;
+  struct: string;
+}) => {
+  const deserialisedKet = ketSerializer.deserializeToDrawingEntities(struct);
+  assert(deserialisedKet);
+  deserialisedKet.drawingEntitiesManager.mergeInto(
+    editor.drawingEntitiesManager,
+  );
+  editor.renderersContainer.update(deserialisedKet.modelChanges);
 };
 
 // TODO: replace after the implementation of the function for processing the structure from the file
-const onOk = ({ struct, fragment }) => {
+const onOk = async ({
+  struct,
+  fragment,
+  onCloseCallback,
+  setIsLoading,
+  dispatch,
+}: {
+  struct: string;
+  fragment: boolean;
+  onCloseCallback: () => void;
+  setIsLoading: (isLoading: boolean) => void;
+  dispatch: Dispatch<AnyAction>;
+}) => {
   if (fragment) {
     console.log('add fragment');
   }
+  const isKet = identifyStructFormat(struct) === SupportedFormat.ket;
   const ketSerializer = new KetSerializer();
   const editor = CoreEditor.provideEditorInstance();
-  const deserialisedKet = ketSerializer.deserializeToDrawingEntities(struct);
-  assert(deserialisedKet);
-  deserialisedKet.drawingEntitiesManager.mergeInto(
-    editor.drawingEntitiesManager,
-  );
-  editor.renderersContainer.update(deserialisedKet.modelChanges);
+  if (isKet) {
+    addToCanvas({ struct, ketSerializer, editor });
+    onCloseCallback();
+    return;
+  }
+  const indigo = IndigoProvider.getIndigo() as StructService;
+  try {
+    setIsLoading(true);
+    const ketStruct = await indigo.convert({
+      struct,
+      output_format: ChemicalMimeType.KET,
+    });
+    addToCanvas({ struct: ketStruct.struct, ketSerializer, editor });
+    onCloseCallback();
+  } catch (error) {
+    const stringError =
+      typeof error === 'string' ? error : JSON.stringify(error);
+    const errorMessage = 'Convert error! ' + stringError;
+    dispatch(openErrorModal(errorMessage));
+    KetcherLogger.error(error);
+  } finally {
+    setIsLoading(false);
+  }
```

Comment:
Later we probably will need to add more formats, so we will need to move this logic somewhere, probably into services as it's done in regular ketcher.